File: graph_loader/src/sphere.cpp
/* The C++ code assumes that you're running it within the context of an Unreal Engine 4 game project. 
The SpawnSpheres() function can be called from anywhere in your code to spawn the spheres.*/

#include "CoreMinimal.h"
#include "Engine/StaticMeshActor.h"
#include "Engine/StaticMesh.h"
#include "Editor/EditorLevelLibrary.h"
#include "Editor/EditorAssetLibrary.h"
#include "Serialization/JsonReader.h"
#include "Serialization/JsonSerializer.h"
#include "Misc/FileHelper.h"

void SpawnSpheres()
{
    FString JsonFilePath = FPaths::ProjectContentDir() + "graphs/path.json";
    FString JsonString;

    // Read the JSON file and load the data
    if (FFileHelper::LoadFileToString(JsonString, *JsonFilePath))
    {
        TSharedPtr<FJsonObject> JsonObject;
        TSharedRef<TJsonReader<TCHAR>> JsonReader = TJsonReaderFactory<TCHAR>::Create(JsonString);

        if (FJsonSerializer::Deserialize(JsonReader, JsonObject))
        {
            // Get a reference to the current level
            UWorld* World = GEditor->GetEditorWorldContext().World();
            ULevel* Level = World->GetCurrentLevel();

            // Define the properties of the sphere
            float SphereRadius = 50.f;

            // Loop through the points and spawn a sphere at each one
            TArray<TSharedPtr<FJsonValue>> Nodes = JsonObject->GetArrayField("nodes");
            for (int32 i = 0; i < Nodes.Num(); ++i)
            {
                FVector Location = FVector(Nodes[i]->AsArray()[0]->AsNumber(),
                                           Nodes[i]->AsArray()[1]->AsNumber(),
                                           Nodes[i]->AsArray()[2]->AsNumber());

                // Create a new sphere actor
                AStaticMeshActor* SphereActor = Cast<AStaticMeshActor>(UEditorLevelLibrary::SpawnActorFromClass(AStaticMeshActor::StaticClass(),
                                                                                                              Location,
                                                                                                              Level->GetActorLocation()));

                // Set the static mesh of the sphere to a built-in sphere mesh
                UStaticMesh* SphereMesh = Cast<UStaticMesh>(UEditorAssetLibrary::LoadAsset("/Engine/BasicShapes/Sphere"));
                SphereActor->GetStaticMeshComponent()->SetStaticMesh(SphereMesh);

                // Set the radius of the sphere
                SphereActor->GetStaticMeshComponent()->SetSphereRadius(SphereRadius);
            }

            // Save the level
            UEditorLevelLibrary::SaveCurrentLevel();
        }
    }
}

File: graph_loader/scripts/sphere.py
import json
import unreal

# Connect to the Unreal Editor
editor = unreal.EditorLevelLibrary()

# Open the JSON file and load the data
with open('../graphs/path.json', 'r') as file:
    data = json.load(file)

# Get a reference to the current level
level = unreal.EditorLevelLibrary.get_editor_world().get_level(0)

# Define the properties of the sphere
sphere_radius = 50

# Loop through the points and spawn a sphere at each one
for node in data['nodes']:
    # Create a new sphere actor
    sphere_actor = unreal.EditorLevelLibrary.spawn_actor_from_class(
        unreal.StaticMeshActor.static_class(),
        unreal.Vector(data['nodes'][node][0], data['nodes'][node][1], data['nodes'][node][2]),
        level.get_actor_location(),
    )

    # Set the static mesh of the sphere to a built-in sphere mesh
    sphere_mesh = unreal.EditorAssetLibrary.load_asset('/Engine/BasicShapes/Sphere')
    sphere_actor.set_editor_property('static_mesh_component', sphere_mesh)

    # Set the radius of the sphere
    sphere_actor.set_editor_property('collision_radius', sphere_radius)

# Save the level
unreal.EditorLevelLibrary.save_current_level()